pandas/_libs/tslibs/offsets.pyx

```diff
@@ -3350,7 +3350,13 @@ cdef class _CustomBusinessMonth(BusinessMixin):
         """
         Define default roll function to be called in apply method.
         """
-        cbday = CustomBusinessDay(n=self.n, normalize=False, **self.kwds)
+        if self.offset:
+            cbday_kwds = self.kwds.copy()
+            cbday_kwds['offset'] = timedelta(0)
+        else:
+            cbday_kwds = self.kwds
+
+        cbday = CustomBusinessDay(n=1, normalize=False, **cbday_kwds)
```

Comment:
can you not just pass kwds & offset here?

Reply:
Likely I didn't fully understanding your suggestion here. Do you by any chance mean passing `**self.kwds` as before? So the idea here is that if _self.offset is not timedelta(0)_, then we always use `CustomBusinessDay(n=1, normalize=False, **offset=timedelta(0)**, ...)` to roll to a business day first, and then separately add the self.offset afterwards. 

The reason is that `.rollforward` (and `.rollback`) only calls `.apply` when if not `.is_on_offset(dt) (code [here](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/offsets.pyx#L590-L616) and [here](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/offsets.pyx#L3227-L3255)), which I think can be illustrated well with the following example:
`
In[33]: import pandas as pd
   ...: print(pd.__version__)
1.3.0.dev0+1628.g32ce31be89
In[34]: cbday_plus_5d = pd.offsets.CustomBusinessDay(n=1, offset=timedelta(days=5))
   ...: # define a CBDay with an extra offset of 5 days
In[35]: print(cbday_plus_5d.rollforward(pd.Timestamp('2021-05-02')))  # expected 2021-05-08
   ...: # 2021-05-02 not a business day, so 'rollforward' makes it 2021-05-03 + extra 5 days
2021-05-08 00:00:00
In[36]: print(cbday_plus_5d.rollforward(pd.Timestamp('2021-05-03')))  # expected 2021-05-08
   ...: # 2021-05-03 is ALREADY a business day, so it doesn't get rolled at all.
2021-05-03 00:00:00
`
So instead, we enforce `offset=timedelta(0)` while constructing `CustomBusinessDay` as follows, which is why I am first modifying the 'offset' parameter:
`
In[42]: cbday_no_offset = pd.offsets.CustomBusinessDay(n=1, **offset=timedelta(0)**)  
In[43]: print(cbday_no_offset.rollforward(pd.Timestamp('2021-05-03')) + timedelta(days=5))  # expected 2021-05-08
2021-05-08 00:00:00
`

Please let me know if I completely missed your question. Would love to hear your thoughts. Thanks!

Reply:
I think what jreback is suggesting is:

```
if self.offset:
    cbday_kwds = self.kwds.copy()
    cbday_kwds['offset'] = timedelta(0)
else:
    cbday_kwds = self.kwds
```

should instead just be:

```
cbday_kwds = self.kwds.copy()
cbday_kwds['offset'] = timedelta(0)
```

Reply:
Ah, I see. I added the check here to prevent unnecessary data copying, because I am guessing `offset=timedelta(0)` is the more common use case and works as expected. What do you think? Definitely happy to change it accordingly.

Reply:
Bumping this thread. Let me know what you think about 1) always making a new copy `cbday_kwds = self.kwds.copy()` vs. 2) only copying when `self.offset` is not `timedelta(0)`.

As a refresher for your convenience, the idea is that we want to first use `CustomBusinessDay(n=1, normalize=False, offset=timedelta(0), ...)` (instead of any non-zero offset) to roll to a business day first, and then separately apply the `self.offset` afterwards in `.apply`.

Reply:
im fine with this either way